zerver/views/scheduled_messages.py

```diff
@@ -9,48 +12,57 @@
 )
 from zerver.lib.exceptions import JsonableError
 from zerver.lib.recipient_parsing import extract_direct_message_recipient_ids, extract_stream_id
-from zerver.lib.request import REQ, RequestNotes, has_request_variables
+from zerver.lib.request import RequestNotes
 from zerver.lib.response import json_success
 from zerver.lib.scheduled_messages import get_undelivered_scheduled_messages
 from zerver.lib.timestamp import timestamp_to_datetime
-from zerver.lib.topic import REQ_topic
-from zerver.lib.validator import check_bool, check_int, check_string_in, to_non_negative_int
+from zerver.lib.typed_endpoint import (
+    ApiParamConfig,
+    OptionalTopic,
+    PathOnly,
+    typed_endpoint,
+    typed_endpoint_without_parameters,
+)
+from zerver.lib.typed_endpoint_validators import check_string_in_validator
 from zerver.models import Message, UserProfile
 
 
-@has_request_variables
+@typed_endpoint_without_parameters
 def fetch_scheduled_messages(request: HttpRequest, user_profile: UserProfile) -> HttpResponse:
     return json_success(
         request, data={"scheduled_messages": get_undelivered_scheduled_messages(user_profile)}
     )
 
 
-@has_request_variables
+@typed_endpoint
 def delete_scheduled_messages(
     request: HttpRequest,
     user_profile: UserProfile,
-    scheduled_message_id: int = REQ(converter=to_non_negative_int, path_only=True),
+    *,
+    scheduled_message_id: PathOnly[NonNegativeInt],
 ) -> HttpResponse:
     delete_scheduled_message(user_profile, scheduled_message_id)
     return json_success(request)
 
 
-@has_request_variables
+@typed_endpoint
 def update_scheduled_message_backend(
     request: HttpRequest,
     user_profile: UserProfile,
-    scheduled_message_id: int = REQ(converter=to_non_negative_int, path_only=True),
-    req_type: str | None = REQ(
-        "type", str_validator=check_string_in(Message.API_RECIPIENT_TYPES), default=None
-    ),
-    req_to: str | None = REQ("to", default=None),
-    topic_name: str | None = REQ_topic(),
-    message_content: str | None = REQ("content", default=None),
-    scheduled_delivery_timestamp: int | None = REQ(json_validator=check_int, default=None),
+    *,
+    scheduled_message_id: PathOnly[NonNegativeInt],
+    req_type: Annotated[
+        Annotated[str, check_string_in_validator(Message.API_RECIPIENT_TYPES)] | None,
+        ApiParamConfig("type"),
+    ] = None,
+    to: Json[int | list[int]] | None = None,
```

Comment:
changed the parameter variable name from `req_to` to `to` since the following was causing a testcase error
```
Annotated[Json[int | list[int]] | None , ApiParamConfig("to")] = None
```
The error was `req_to` needs to be "application/json" despite it being json and the error not appearing when ApiParamConfig is not being used.

Seems to be an error very specific to `Json[int | list[int]] | None` because similar implementations of `Json[int] | None` were not flagged as an error.

I wasn't able to figure out exactly why this was happening, therefore resorted to not using `ApiParamConfig`.

Reply:
This is a bit mysterious, but the end result seems OK.